lib/request.coffee

```diff
@@ -322,4 +306,53 @@ module.exports = getRequest = ({
 	# (or resolve to) a new response, or throw/reject.
 	###
 
+	###*
+	# @summary Refresh token on user request
+	# @function
+	# @public
+	#
+	# @description
+	# This function automatically refresh token on user request.
+	# The token is refreshed even if it is not expired yet.
+	# Error is thrown if user is not authorized.
+	#
+	# @param {String} options.url - relative url
+	#
+	# @returns {String} token - new token
+	#
+	# @example
+	# request.refreshToken
+	# 	baseUrl: 'https://api.resin.io'
+	###
+
+	exports.refreshToken = (options = {}) ->
+
+		{baseUrl} = options
+
+		# Only refresh if we have resin-auth
+		return if not (auth?)
+
+		exports.send
+			url: '/whoami'
+			baseUrl: baseUrl
+			refreshToken: false
+
+		# At this point we're sure there is a saved token,
+		# however the fact that /whoami returns 401 allows
+		# us to safely assume the token is expired
+
+		.catch
+			code: 'ResinRequestError'
+			statusCode: 401
+		, ->
+			console.log('error catched')
+			return auth.getKey().tap(auth.removeKey).then (key) ->
+				console.log('resin expired token')
+				throw new errors.ResinExpiredToken(key)
+
+		.get('body')
+		.then(auth.setKey)
+		.then ->
+			auth.getKey()
```

Comment:
We're using Bluebird which comes with a bunch of [nice helpers](http://bluebirdjs.com/docs/api-reference.html) for this kind of thing. Here we can use `tap` to simplify this and avoid having to reload the key:

```coffee
.get('body')
.tap(auth.setKey) # Same as .then, but keeps the previous return value
``` 

That lets you skip this last `.then` entirely.